[Concurrency] Add an environment variable to validate unchecked continuation usage.

mikeash opened the pull request:
When enabled, we track all active unchecked continuations in a global set, and fatal error if one is called twice.

This commit also removes `swift_Concurrency_fatalError` in favor of exporting `swift::fatalError` from libswiftCore as SPI so that libswift_Concurrency can use it. This gives us better error messages and crash reporting for fatal errors in libswift_Concurrency.

rdar://97390481

mikeash commented:
@kateinoigakukun I saw you added `swift_Concurrency_fatalError`, will there be any problem with removing that in favor of exporting `swift::fatalError` as SPI and calling that?

kateinoigakukun commented:
@mikeash As Doug told me in my PR, `swift::fatalError` is not exported by older `swiftCore`, so simple linkage against it in Concurrency runtime would break back-deploy, I think.
https://github.com/apple/swift/pull/37084#pullrequestreview-646122992

Unfortunately, this breakage would be happened by `concurrencyValidateUncheckedContinuations` also.

If we want to add dependencies on newer `swiftCore`'s symbols, we need to guard the use of newly exported symbols like [here](https://github.com/apple/swift/blob/c019b240762f9297c8718984ba74dd17388dc080/stdlib/public/Concurrency/DispatchGlobalExecutor.inc#L89), or add the impls of those symbols in back-deploy library.

CC: @DougGregor

mikeash committed:
[Concurrency] Add an environment variable to validate unchecked continuation usage.

When enabled, we track all active unchecked continuations in a global set, and fatal error if one is called twice.

rdar://97390481

mikeash commented:
The fatal error stuff was causing too much trouble and didn't belong here. I've removed that and rebased the rest of the PR on top of `main`.

mikeash commented:
@swift-ci please test

mikeash commented:
I didn't realize this PR was from a branch in the main repo! I rebased and made a new PR based on my own fork here: https://github.com/apple/swift/pull/61076
